src/Flysystem/Filesystem.php

```diff
@@ -33,4 +40,61 @@ public function has(string $location): bool
 
         return strlen($path) === 0 ? false : ($this->getAdapter()->fileExists($path) || $this->getAdapter()->directoryExists($path));
     }
+
+    /**
+     * Check if a directory is empty
+     */
+    public function isEmpty(string $location): bool
+    {
+        // listContents() uses generators, so we can start the iterator and return false if there's a single item.
+        // In most cases this will be orders of magnitude faster than checking if $this->listContents($location)->toArray() is empty.
+        foreach ($this->listContents($location) as $item) {
+            return false;
+        }
+        return true;
+    }
+
+    public function listContentsByGlob(string $folder, string $fileGlob, bool $deep = FilesystemReader::LIST_SHALLOW): DirectoryListing
+    {
+        $path = $this->pathNormalizer->normalizePath($folder);
+        if ($this->adapter instanceof GlobContentLister) {
+            $listing = $this->adapter->listContentsByGlob($path, $fileGlob, $deep);
+            return new DirectoryListing($this->pipeListing($folder, $deep, $listing));
+        }
+
+        // If the adapter isn't a GlobContentLister, we have to filter by glob after fetching the content.
+        // Use regex to match the found file paths against the glob pattern.
+        // Regex starts with folder location
+        $globRegex = '#^' . rtrim(preg_quote($path, '#'), '/') . '/';
+        // Allow any number of subdirs if checking recursively
+        if ($deep) {
+            $globRegex .= '.*/?';
+        }
+        // Convert original file glob to regex and make sure any period is escaped
+        // Asset file names shouldn't contain any special regex characters other than period so this should be safe.
+        $globRegex .= ltrim(Glob::toRegex($fileGlob, false, true, '#'), '#^');
+        $listing = $this->adapter->listContents($path, $deep);
+        return new DirectoryListing($this->pipeListing(
+            $folder,
+            $deep,
+            $listing,
+            fn (StorageAttributes $item): bool => preg_match($globRegex, $item->path())
+        ));
```

Comment:
We could alternatively have a `public function canListByGlob(): bool` which returns `false` if the adapter isn't a `GlobContentLister` - but that would require one of the following:

1. Have two different interfaces instead of the one
1. Do a `method_exists()` check after the `instanceof` check
1. Call `canListByGlob()` even though it's not in the type we've checked (ew)
1. ??? I feel like there's something else but my brain won't tell me what.

All in all I think it's easier and probably cleaner to just always call `listContentsByGlob()` regardless of what the adapter is, and then if the adapter can't glob, we just do glob pattern matching post-fetch (i.e. what we're doing here)